drivers/regulator/regulator_da1469x.c

```diff
@@ -0,0 +1,451 @@
+
+LOG_MODULE_REGISTER(regulator_da1469x, CONFIG_REGULATOR_LOG_LEVEL);
+
+/* Rail should be powered during sleep, mode applies to VDD, V18, V18P */
+#define DA1469X_MODE_SLEEP_LDO					0x01
+/* Settings valid for VDD, V14, V18, V18P */
+#define DA1469X_MODE_DCDC_HIGH_BATTERY_VOLTAGE_ON		0x02
+#define DA1469X_MODE_DCDC_LOW_BATTERY_VOLTAGE_ON		0x04
+/* Settings valid for V30 */
+#define DA1469X_MODE_LDO_VBUS_ON				0x08
+#define DA1469X_MODE_LDO_VBAT_ON				0x10
+#define DA1469X_MODE_V30_VBAT_CLAMP_ON				0x20
+
+#define DCDC_REG_FIELD_MSK(dcdc, field) \
+	DCDC_DCDC_ ## dcdc ## _REG_DCDC_ ## dcdc ## _ ## field ## _Msk
+#define DCDC_REG_FIELD_POS(dcdc, field) \
+	DCDC_DCDC_ ## dcdc ## _REG_DCDC_ ## dcdc ## _ ## field ## _Pos
+
+#define DCDC_REG_GET(dcdc, field) \
+	((DCDC->DCDC_ ## dcdc ## _REG & DCDC_REG_FIELD_MSK(dcdc, field)) >> \
+	DCDC_REG_FIELD_POS(dcdc, field))
+
+#define DCDC_REG_SET(dcdc, field, val) \
+	DCDC->DCDC_ ## dcdc ## _REG = \
+	((DCDC->DCDC_ ## dcdc ## _REG & ~DCDC_REG_FIELD_MSK(dcdc, field)) | \
+	((val) << DCDC_REG_FIELD_POS(dcdc, field)))
+
+#define POWER_REG_FIELD_MSK(field) \
+	CRG_TOP_POWER_CTRL_REG_ ## field ## _Msk
+#define POWER_REG_FIELD_POS(field) \
+	CRG_TOP_POWER_CTRL_REG_ ## field ## _Pos
+
+#define POWER_REG_GET(field) \
+	((CRG_TOP->POWER_CTRL_REG & POWER_REG_FIELD_MSK(field)) >> POWER_REG_FIELD_POS(field))
+
+#define POWER_REG_SET(field, val) \
+	CRG_TOP->POWER_CTRL_REG = \
+	((CRG_TOP->POWER_CTRL_REG & ~POWER_REG_FIELD_MSK(field)) | \
+	((val) << POWER_REG_FIELD_POS(field)))
+
+static const uint32_t voltages_vdd_clamp[] = {
+	1037000, 1005000, 978000, 946000,
+	1120000, 1089000, 1058000, 1030000,
+	952000, 918000, 889000, 861000,
+	862000, 828000, 798000, 706000
+};
+
+static const uint32_t voltages_vdd[] = {
+	900000, 1000000, 1100000, 1200000
+};
+
+static const uint32_t voltages_vdd_sleep[] = {
+	750000, 800000, 850000, 900000,
+};
+
+static const uint32_t voltages_v14[] = {
+	1200000, 1250000, 1300000, 1350000,
+	1400000, 1450000, 1500000, 1550000
+};
+
+static const uint32_t voltages_v30[] = {
+	3000000, 3300000
+};
+
+static const uint32_t voltages_v18[] = {
+	1200000, 1800000
+};
+
+static const uint32_t voltages_v18p[] = { 1800000 };
+
+enum da1469x_rail {
+	VDD_CLAMP,
+	VDD_SLEEP,
+	VDD,
+	V14,
+	V18,
+	V18P,
+	V30,
+};
+
+struct regulator_da1469x_config {
+	struct regulator_common_config common;
+	const uint32_t *voltages;
+	const uint8_t voltages_count;
+	enum da1469x_rail rail;
+	uint8_t flags;
+};
+
+struct regulator_da1469x_data {
+	struct regulator_common_data common;
+};
+
+static int regulator_da1469x_enable(const struct device *dev)
+{
+	const struct regulator_da1469x_config *config = dev->config;
+
+	switch (config->rail) {
+	case VDD:
+		POWER_REG_SET(LDO_CORE_ENABLE, 1);
+		DCDC_REG_SET(VDD, ENABLE_HV,
+			     !!(config->flags & DA1469X_MODE_DCDC_HIGH_BATTERY_VOLTAGE_ON));
+		DCDC_REG_SET(VDD, ENABLE_LV,
+			     !!(config->flags & DA1469X_MODE_DCDC_LOW_BATTERY_VOLTAGE_ON));
```

Comment:
These could be combined into a single write.

Reply:
after rework registers modified in one write